File: solutions/Leetcode_968/leetcode_968.py
class Solution1:
    def min_camera_cover(self, root):
        self._num_camera = 0
        def dfs(node):
            if not node:
                return 2
            left, right = dfs(node.left), dfs(node.right)
            if left == 0 or right == 0:
                self._num_camera += 1
                return 1
            return 2 if left == 1 or right == 1 else 0
        return (dfs(root) == 0) + self._num_camera


class SolutionMINE:
    def min_camera_cover(self, root):
        self._num_camera = 0
        dummy = TreeNode(0, left=root, right=None)
        self.helper(dummy)
        return self._num_camera

    def helper(self, node):
        left = self.helper(node.left) if node.left else 0
        right = self.helper(node.right) if node.right else 0
        if left == 2 or right == 2:
            self._num_camera += 1
        if left == right == 0:
            return 2
        return max(left, right) - 1

File: solutions/Leetcode_968/leetcode_968.cc
class Solution1
{
public:
    int minCameraCover(TreeNode* root)
    {
        return (dfs(root) == 0 ? 1 : 0) + num_cameras;
    }

private:
    int num_cameras = 0;

    int dfs(TreeNode* node)
    {
        if (node == nullptr)
            return 2;
        int left = dfs(node->left);
        int right = dfs(node->right);
        if (left == 0 || right == 0)
        {
            num_cameras++;
            return 1;
        }
        return left == 1 || right == 1 ? 2 : 0;
    }
};